File: dbflow/src/main/java/com/raizlabs/android/dbflow/structure/database/DatabaseWrapper.java
package com.raizlabs.android.dbflow.structure.database;

import android.content.ContentValues;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Description: Provides a base implementation that wraps a database, so other database engines potentially can
 * be used.
 */
public interface DatabaseWrapper {

    void execSQL(@NonNull String query);

    void beginTransaction();

    void setTransactionSuccessful();

    void endTransaction();

    int getVersion();

    @NonNull
    DatabaseStatement compileStatement(@NonNull String rawQuery);

    @NonNull
    FlowCursor rawQuery(@NonNull String query, @Nullable String[] selectionArgs);

    long updateWithOnConflict(@NonNull String tableName,
                              @NonNull ContentValues contentValues,
                              @Nullable String where,
                              @Nullable String[] whereArgs, int conflictAlgorithm);

    long insertWithOnConflict(@NonNull String tableName,
                              @Nullable String nullColumnHack,
                              @NonNull ContentValues values,
                              int sqLiteDatabaseAlgorithmInt);

    @NonNull
    FlowCursor query(@NonNull String tableName, @Nullable String[] columns, @Nullable String selection,
                     @Nullable String[] selectionArgs, @Nullable String groupBy,
                     @Nullable String having, @Nullable String orderBy);

    int delete(@NonNull String tableName, @Nullable String whereClause, @Nullable String[] whereArgs);
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/database/DatabaseWrapper.kt
package com.raizlabs.android.dbflow.database

import android.content.ContentValues

/**
 * Description: Provides a base implementation that wraps a database, so other databaseForTable engines potentially can
 * be used.
 */
interface DatabaseWrapper {

    val version: Int

    fun execSQL(query: String)

    fun beginTransaction()

    fun setTransactionSuccessful()

    fun endTransaction()

    fun compileStatement(rawQuery: String): DatabaseStatement

    fun rawQuery(query: String, selectionArgs: Array<String>?): FlowCursor

    fun updateWithOnConflict(tableName: String,
                             contentValues: ContentValues,
                             where: String?,
                             whereArgs: Array<String>?, conflictAlgorithm: Int): Long

    fun insertWithOnConflict(tableName: String,
                             nullColumnHack: String?,
                             values: ContentValues,
                             sqLiteDatabaseAlgorithmInt: Int): Long

    fun query(tableName: String, columns: Array<String>?, selection: String?,
              selectionArgs: Array<String>?, groupBy: String?,
              having: String?, orderBy: String?): FlowCursor

    fun delete(tableName: String, whereClause: String?, whereArgs: Array<String>?): Int
}
